NGINX and Let's Encrypt tutorial

smashwilson opened the pull request:
I'm kind of merging [@rgbkrk's writeup](https://getcarina.com/blog/weekly-news-docker-sock-letsencrypt/#let-39-s-encrypt-with-free-certificates) with [mine](https://getcarina.com/blog/push-button-lets-encrypt/) to create a comprehensive Let's Encrypt-on-nginx guide. I'm adapting things somewhat to be usable in multi-segment clusters as well.

- [x] Use [build affinities](https://docs.docker.com/swarm/scheduler/filter/) to build the NGINX and cron container images on the same segment as the data container.
- [x] Actually test the cron container with a staging setup.
  - [x] Include docker in the cron container (whoops)
  - [x] webrootauth does not exist or is not a directory
- [x] Include rate-limit errors in the troubleshooting section.

Fixes #652.

everett-toews commented:
You could also let the nginx container land where it may and use a container name affinity to put the cron container next to it with `--env affinity:container==lets-nginx`

smashwilson commented:
@everett-toews: I'm not using lets-nginx for this. I wanted to keep our tutorials using official container images as much as possible, and be clear what's actually going on - lets-nginx is very much a "launch it and it hides all the details from you" solution. I'll likely link my article (and Kyle's) in the "resources" section.

For affinities, I'm using a data volume container to hold the certs and using affinities to it (mostly implicitly via `--volumes-from`) to keep things colocated on the same segment. That also protects us from having to deal with the weirdness you discovered with `docker volume`.

smashwilson commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,295 @@
+
+    ```bash
+    $ docker build \
+      --build-arg affinity:container==letsencrypt-data \
````
The Swarm docs seem to imply that this won't work, but it certainly seems to for me?

smashwilson commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,295 @@
+
+    Notice that running this container will automatically accept the Let's Encrypt [terms of service](https://letsencrypt.org/repository/) on your behalf. Be sure that you're okay with accepting them before you run this!
+
+    The `--server` parameter below will issue certificates from Let's Encrypt's *staging* server, which means that these certificates won't be browser-valid yet. Let's Encrypt rate-limits certificate issuance to [five certificates per domain per seven days](https://community.letsencrypt.org/t/public-beta-rate-limits/4772/3), so it's a good idea to use staging certificates until you're confident t ...
```
I'm defaulting all of these commands to get staging certificates to be copy-and-paste friendly without burning through your domain's certs while you're working out the kinks. On the other hand, this does mean that running through verbatim won't actually get you that A+ and green lock. Thoughts?

everett-toews replied:
Works for me.

everett-toews replied:
This link is okay too.

everett-toews replied:
You know, after running into the problem of having to delete the DVC maybe it's better to just go the production cert route from the get go. I think you'd still want to include the caveat above but rework it to say use `--server https://acme-staging.api.letsencrypt.org/directory` if you're just testing...delete the DVC if you do...yada yada yada.

Thoughts?

smashwilson commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,295 @@
+      --detach \
+      --name my-cron \
+      --volume /var/run/docker.sock:/var/run/docker.sock \
+      --env affinity:container==letsencrypt-data \
```
You actually do need the affinity here, even though `reissue` runs Docker commands. Because I'm mounting the *segment's* Docker sock (rather than fussing with `DOCKER_HOST` for the swarm cluster) the Docker commands in that script will launch containers on the same segment, so they'll break unless this container has affinity.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,295 @@
+    A data volume container is a container that exists only to house a Docker volume. It's usually implemented as a container whose process terminates immediately. To learn more, read the [data volume container tutorial]({{ site.baseurl }}/docs/tutorials/data-volume-containers/).
+
+    ```bash
+    $ docker run \
````
In [Create a DVC](https://getcarina.com/docs/tutorials/data-volume-containers/#create-a-dvc) I recommend `docker create`. Is there any benefit to actually running it with `docker run`?

smashwilson replied:
Nope! Good catch.

smashwilson replied:
Hah. Actually I'm going to use this to run a `mkdir` so that the webroot directory exists within the container. letencrypt dies on reauth if it doesn't and this lets us save a step creating it.

smashwilson commented:
Testing out the cron container with a staging cert and a 15 minute renewal interval now.

smashwilson commented:
Looks like I have the reissue job working:

```
crond: USER root pid   6 cmd run-parts /etc/periodic/hourly
IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at
   /etc/letsencrypt/live/staging.farts.horse/fullchain.pem. Your cert
   will expire on 2016-04-28. To obtain a new version of the
   certificate in the future, simply run Let's Encrypt again.
 - If you like Let's Encrypt, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le

staging-nginx
```

Phew. Now to pull out reusable sections into `_include`.

smashwilson commented:
This tutorial is already pretty long. Setting up Cloud Files backups would make it even longer, and I'm not sure we should be advising people to put TLS private keys into Cloud Files anyway.

I would say that backups should be their own tutorial and we could link it here as a "next step." Make sense?

smashwilson commented:
Alright, actually ready for review now :sparkles:

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+
+1. Create a data volume container to hold the Let's Encrypt certificates and account information.
+
+    A data volume container is a container that exists only to house a Docker volume. It's usually implemented as a container whose process terminates immediately. To learn more, read the [data volume container tutorial]({{ site.baseurl }}/docs/tutorials/data-volume-containers/).
```
We generally try to avoid linking to things in the body of tutorials unless is somewhere we specifically want the user to go right now. Best to move this link to the Resources section.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+  - nginx
+---
+
+This tutorial describes acquiring free TLS certificates from [Let's Encrypt](https://letsencrypt.org/) and using them to secure a web site served using [NGINX](http://nginx.org/en/docs/). This encrypts all traffic to and from your site with a certificate that's trusted by your browser and that provides strong enough encryption to get an "A+" rating from [SSL Labs](https://www.ssllabs.com/ssltest/ ...
```
We generally try to avoid linking to things in the body of tutorials unless is somewhere we specifically want the user to go right now. I think the Let's Encrypt link is okay because it's central to the tutorial. The NGINX link can be removed. The SSL Labs link is used at the end of the tutorial so it isn't necessary here too.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+      --entrypoint /bin/mkdir \
+      quay.io/letsencrypt/letsencrypt \
+      -p /etc/letsencrypt/webrootauth/
+    ```
````
Explain that no output is expected here.

smashwilson replied:
Done :+1:

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+      -p /etc/letsencrypt/webrootauth/
+    ```
+
+1. Generate [strong Diffie-Hellman parameters](https://raymii.org/s/tutorials/Strong_SSL_Security_On_nginx.html#Forward_Secrecy_&_Diffie_Hellman_Ephemeral_Parameters) and store them within your data volume container. These will prevent NGINX from using weaker parameters while negotiating the initial TLS connection, and are necessary to reach that "A+" rating on SSL labs.
````
This link can move to resources.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+      --rm --interactive --tty \
+      --volumes-from letsencrypt-data \
+      nginx \
+      openssl dhparam -out /etc/letsencrypt/dhparams.pem 2048
```
Include a sample of the output after the command.

```
  Generating DH parameters, 2048 bit long safe prime, generator 2
This is going to take a long time
...
```

smashwilson replied:
Got it.

(It's hard to do elipses on something that's already marking progress with elipses...)

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+    Even if your cluster has multiple segments, all of your containers will be running on the same segment as the data volume container you just created. To find its public IP address, run:
+
+    ```bash
+    $ docker inspect --format "{{ "{{ .Node.IP "}}}}" letsencrypt-data
````
Should be `docker inspect --format "{{ .Node.IP }}" letsencrypt-data`

everett-toews replied:
Include output of this command. It's okay to use a specific address, as long as that address gets used everywhere else appropriately in the tutorial.

smashwilson replied:
> Should be docker inspect --format "{{ .Node.IP }}" letsencrypt-data

This is how you produce exactly that output through Jekyll. Liquid doesn't let you escape with any variant of `\{{ ... }}`; if you want to produce a literal `{{ .. }}` you need to interpolate a string in. I verified with the local client (it took a little trial-and-error to get it right).

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+    **Mac OS X and Linux**
+
+    ```bash
+    $ dig +short <myDomain>
````
Include output of this command and mention that it should match the output of the `docker inspect` command.

smashwilson replied:
:+1: Done.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+    **Windows**
+
+    ```powershell
+    > nslookup <myDomain>
````
Output please.

smashwilson replied:
Actually took a trip downstairs to the Windows box verify this one :wink:

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+
+1. Run a Let's Encrypt container to issue a new certificate. The Let's Encrypt client must be able to bind to ports 80 and 443 on the segment.
+
+    Notice that running this container will automatically accept the Let's Encrypt [terms of service](https://letsencrypt.org/repository/) on your behalf. Be sure that you're okay with accepting them before you run this!
```
This link is good stuff.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+      --agree-tos
+    ```
+
+    If all goes well, you'll see output like this:
````
Huh. I ran it without the `--server` param immediately after I did with it. I got this output. 

```
User chose to cancel the operation and may reinvoke the client.
IMPORTANT NOTES:
 - If you lose your account credentials, you can recover through
   e-mails sent to everett.toews@gmail.com.
 - Your account credentials have been saved in your Let's Encrypt
   configuration directory at /etc/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Let's
   Encrypt so making regular backups of this folder is ideal.
```

Is `User chose to cancel the operation and may reinvoke the client.` to be expected?

everett-toews replied:
After deleting the DVC and rerunning letsencrypt, the output appeared as below.

smashwilson replied:
This now references the full "getting a production certificate" section.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+    <!doctype html>
+    <html lang="en-us">
+        <head>
+            <title>Let's Encrypted</title>
```
Spelled "Encrypted" on purpose?

smashwilson replied:
Yah, just a bit of wordplay.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+    </html>
+    ```
+
+    Now create the NGINX configuration itself. Use the [Mozilla SSL configuration generator](https://mozilla.github.io/server-side-tls/ssl-config-generator/) as a starting point to create a `default.conf` that uses up-to-date, secure settings, and adapt it to meet your needs. If you use this example, remember to replace `<myDomain>` with your domain name.
````
Good link. I didn't know this existed!

everett-toews replied:
`...replace `<myDomain>` **everywhere** with your domain name.`

Need to stress everywhere. I totally missed some and borked by first deploy.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+      my-nginx
+    ```
+
+    Visit your domain in your browser. If it's still a staging certificate, you'll get a security warning, but your site is being served over TLS! If you're unable to connect or you see an error message, instead, check the [troubleshooting section](#troubleshooting) for ways to diagnose your problem.
````
Going the production server route, at this point I get an `NET::ERR_CERT_AUTHORITY_INVALID`.

I had to go back and completely remove my data volume container. You'll need to document that the DVC needs to be removed before going with a production cert.

smashwilson replied:
> I had to go back and completely remove my data volume container. You'll need to document that the DVC needs to be removed before going with a production cert.

Hrm - this is what I get for doing most of my infra testing with staging certs. Yeah, I guess we need either a DVC removal or some kind of `docker run --rm -it --volumes-from letsencrypt-data alpine rm -r /etc/letsencrypt /var/lib/letsencrypt/` deal. Going that route would keep you from needing to recreate _all_ of the containers that mount its volume.

Where do you think that belongs? Just after the last step? As its own step, just as an afternote, or as its own full section?

everett-toews replied:
Good idea. In the main body of the tutorial just do staging. At the end of it have a completely new section for production just before the Troubleshooting section. The caveat in the main body can reference the production section easily.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
```diff
@@ -0,0 +1,329 @@
+
+1. Create a script that will reissue your certificates. Let's Encrypt issues certificates that expire in a relatively short 90-day period to mitigate the risk of compromised credentials and to encourage you to automate re-issuance.
+
+    Write `reissue`, substituting your domain and email address. Again, this example uses `--server` to target the Let's Encrypt staging environment; omit the `--server` argument and its parameter for your production infrastructure.
```
Create a file named `reissue` using the contents below and substitute your domain and email address.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+    docker kill --signal=HUP my-nginx
+    ```
+
+1. Write a `Dockerfile.cron` that uses `crond` to invoke this script once every month. This example uses [Alpine](https://hub.docker.com/_/alpine/) as a base image to keep the image small. Conveniently, Alpine includes an [`/etc/periodic` directory](http://wiki.alpinelinux.org/wiki/Alpine_Linux:FAQ#My_cron_jobs_don.27t_run.3F) that can be used to easily run scripts at a variety of intervals. All  ...
````
Create a file named `Dockerfile.cron` using the contents below. This Dockerfile uses crond to invoke the script once every month.

everett-toews replied:
Remove both links.

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+1. Put it all together with a `Dockerfile.nginx`:
+
+    ```Dockerfile
+    FROM nginx
````
If you're using `alpine:3.3` below, want to use it here too like [this](http://gliderlabs.viewdocs.io/docker-alpine/usage/#user-content-disabling-cache)?

smashwilson replied:
Hmm, I was defaulting to using the official Docker images when possible. I went with alpine for the cron container because there isn't an official cron container.

everett-toews replied:
SGTM

everett-toews commented on _tutorials/2016-01-26-securing-nginx-with-lets-encrypt.md:
````diff
@@ -0,0 +1,329 @@
+      .
+    ```
+
+1. Run your cron job in a dedicated cron container. Remember to mount the Docker socket from the host so that the `reissue` script's `docker` commands will work, and to specify an affinity to your data container to run on the correct segment:
````
No need to have "Remember to". Just start with "Mount the..."

everett-toews commented:
:tada: https://test.phymata.com/
:tada: https://www.ssllabs.com/ssltest/analyze.html?d=test.phymata.com

This is a great tutorial. It's definitely the most complex one we've had so far. Thanks @smashwilson!

smashwilson commented:
Sweet - thanks for the review! I'll get this fixed up and add that "switching to production" section on Monday.

rgbkrk commented:
This is awesome!

rgbkrk commented:
Side note: because my name was in the link, GitHub didn't notify me.

smashwilson commented:
@everett-toews Alright, I think I addressed all of your suggestions. I also split the "getting a production certificate" bit into its own section rather than do that weird "re-run some of this without this parameter" thing.

Let me know if I missed anything or if there's anything weird in the new bit.

everett-toews commented:
:+1: 

@kmsholcomb Can you please give this a review this week?

kmsholcomb commented:
Yes, will do!

kmsholcomb commented:
Hi @smashwilson and @everett-toews: The attached Word doc (and a PDF copy of it, in case you don't like Word) contain my suggested copy edits plus suggestions for breaking the long procedure up into smaller ones. Please let me know if you have any questions. Thanks!

[Securing NGINX with Let's Encrypt_editsKH.pdf](https://github.com/getcarina/getcarina.com/files/114555/Securing.NGINX.with.Let.s.Encrypt_editsKH.pdf)
[Securing NGINX with Let's Encrypt_editsKH.docx](https://github.com/getcarina/getcarina.com/files/114556/Securing.NGINX.with.Let.s.Encrypt_editsKH.docx)

smashwilson commented:
Alright, final edits are in. @kmsholcomb, I followed your lead and broke the long sequence of steps into discrete section - it's much nicer and more readable this way. I did balance the steps a little differently (it made more sense to me to do it by container).

smashwilson commented:
@everett-toews, @kmsholcomb: Any final comments, or shall I merge tomorrow?

kmsholcomb commented:
Hey @smashwilson: It looks great! I'm glad that you improved upon my suggestions, which were just stabs at chunking the content. Certainly what makes sense to you will make sense to our users. :+1:

rgbkrk commented:
Might as well update the date and :shipit:.

smashwilson commented:
Aye aye! :shipit: in progress...
